students/prgrover/lesson06/mailroom.py

```diff
@@ -0,0 +1,184 @@
+
+    ''')
+
+
+def get_donor():
+    """
+    Prompts user to enter a donor's name.
+
+    Args: None
+    """
+
+    donor_name = input("Please enter the full name of a donor. You can also enter list to see all current donors or 'e' to exit: ")
+    if donor_name == "e":
+        menu_selection(main_prompt, main_dispatch)
+    elif donor_name == "list":
+        print(list_donors())
+        menu_selection(main_prompt, main_dispatch)
+    else:
+        return donor_name
+
+
+def list_donors():
+    """
+    Returns a list of all donors.
+
+    Args: None
+    """
+    donor_list = "\n".join(donor_dict.keys())
+    return donor_list
+
+
+def add_new_donor(donor_name):
+    """
+    Adds a new donor and donation to the dictionary.
+
+    Args:
+    donor_name:  The donor's name
+    donotion_amount:  The donoation amount
+    """
+    donor_dict[donor_name] = []
+
+
+def get_donation():
+    """
+    Prompts the user for a donoation amount and returns the value.
+
+    Args: None
+    """
+    while True:
+        donation_amount = input("Please enter the donation amount or 'e' to exit: ")
+        if donation_amount == "e":
+            menu_selection(main_prompt, main_dispatch)
+        else:
+            try:
+                donation_amount = float(donation_amount)
+            except ValueError:
+                print("Invalid donation amount. Please try again.\n")
+            else:
+                return donation_amount
+
+
+def add_donation(donor_name, donation_amount):
+    """
+    Adds a new donation amount for new and existing donors.
+
+    Args:
+    donor_name:  A donor
+    donation_amount:  Donation amount for the dono
+    """
+    if donor_name in donor_dict:
+        donor_dict[donor_name].append(donation_amount)
+    else:
+        add_new_donor(donor_name)
+        donor_dict[donor_name].append(donation_amount)
+
+
+def send_thankyou():
+    """
+    Provides the following functionality:
+    1. Lists all donors
+    2. Adds new donors
+    3. Enters new donations
+    4. Sends a thank you email
+
+    Args: None
+    """
+    donor = get_donor()
+    donation = get_donation()
+    add_donation(donor, donation)
+    print(email.format(donor, donation))
+    return
+
+
+def create_report():
+    """
+    Prints a report of all donors sorted by total donations descending.
+
+    Args: None
+    """
+    print ("{:30s} | {:11s} | {:9s} | {:12s}".format("Donor Name", "Total Given", "Num donations", "Average donation"))
+    print ("-" * 79)
+
+    sorted_donor_dict = OrderedDict(sorted(donor_dict.items(), key=lambda i: sum(i[1]), reverse=True))
+    
+    for (donor, donations) in sorted_donor_dict.items():
+        donor_name = donor
+        total_given = sum(donations)
+        num_donations = len(donations)
+        average_donation = (total_given / num_donations)
+        print ("{:30s}  ${:11.2f}   {:12d}    ${:12.2f}".format(donor_name, total_given, num_donations, average_donation))
+
```

Comment:
You could have this function return the report as a string to be printed, rather than printing as a side effect of the function, to make it testable